flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/SubstringIndexFunction.java

```diff
@@ -0,0 +1,155 @@
+        if (expr == null || delim == null || count == null) {
+            return null;
+        }
+
+        if (count == 0) {
+            return BinaryStringData.EMPTY_UTF8;
+        }
+        String str = expr.toString();
+        if (str.isEmpty()) {
+            return BinaryStringData.EMPTY_UTF8;
+        }
+        String delimiter = delim.toString();
+        if (delimiter.isEmpty()) {
+            return BinaryStringData.EMPTY_UTF8;
+        }
+
+        try {
+            if (count > 0) {
+                // get index of the count-th occurrences of delimiter
+                int idx = StringUtils.ordinalIndexOf(str, delimiter, count);
+                if (idx != StringUtils.INDEX_NOT_FOUND) {
+                    return StringData.fromString(str.substring(0, idx));
+                }
+            } else {
+                // get index of the last -count-th occurrences of delimiter
+                int idx = StringUtils.lastOrdinalIndexOf(str, delimiter, -count);
+                if (idx != StringUtils.INDEX_NOT_FOUND) {
+                    return StringData.fromString(str.substring(idx + delimiter.length()));
+                }
+            }
+        } catch (Throwable t) {
+            throw new FlinkRuntimeException(t);
+        }
+
+        // can not find enough delimiter
+        return expr;
+    }
+
+    public @Nullable byte[] eval(
+            @Nullable byte[] expr, @Nullable byte[] delim, @Nullable Integer count) {
+        if (expr == null || delim == null || count == null) {
+            return null;
+        }
+        if (expr.length == 0 || delim.length == 0 || count == 0) {
+            return new byte[] {};
+        }
+
+        if (count > 0) {
+            int idx = -1;
+            while (count > 0) {
+                // get index of the next occurrence of delim
+                idx = find(expr, idx + 1, delim);
+                if (idx >= 0) {
+                    --count;
+                } else {
+                    // can not find enough delim
+                    return expr;
+                }
+            }
+            if (idx == 0) {
+                return new byte[] {};
+            }
+            return Arrays.copyOfRange(expr, 0, idx);
+
+        } else {
+            int idx = expr.length - delim.length + 1;
+            while (count < 0) {
+                // get index of the previous occurrence of delim
+                idx = rfind(expr, idx - 1, delim);
+                if (idx >= 0) {
+                    ++count;
+                } else {
+                    // can not find enough delim
+                    return expr;
+                }
+            }
+            if (idx + delim.length == expr.length) {
+                return new byte[] {};
+            }
+            return Arrays.copyOfRange(expr, idx + delim.length, expr.length);
+        }
+    }
+
+    private int find(byte[] str, int beginIdx, byte[] target) {
+        final int endIdx = str.length - target.length;
```

Comment:
Do we should check the str.length > target.length?

Reply:
I believe it is unnecessary, as the current code would return `-1` directly in this case, indicating no matching substring found, which aligns with the expected behavior.

Reply:
BTW, I do find a bug in the next row, `beginIndex < endIndex` should be `beginIndex <= endIndex`. Corresponding test cases have been added.